src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpContext.cs

```diff
@@ -260,6 +237,14 @@ private IISAwaitable DoFlushAsync()
 
         public Task WriteAsync(ArraySegment<byte> data, CancellationToken cancellationToken = default(CancellationToken))
         {
+            // To avoid concurrent reading and writing, if we have a pending read,
+            // we must cancel it. TODO check if we need to lock here. I don't think we do because cancel_io will
+            // noop if there is no data to write, and OnAsyncCompletion should only be triggered once per read/write.
+            if (!_wasUpgraded && !_operation.IsCompleted)
+            {
+                var hr = NativeMethods.http_cancel_io(_pInProcessHandler);
```

Comment:
This seems very unstable. How do your read operations distinguish between a cancellation and real IO exception? This would need a lot of stress testing.

While we understand the benefits of pipelines for writes, are we clear on what value they provide for reads? I'm not aware of any.

Reply:
Setting aside the potential issue of distinguishing canceled reads and IO errors, I think there might be a race condition here. What happens if http_cancel_io is called right before the call to http_read_request_bytes? Does the http_read_request_bytes call still get canceled?

Reply:
Yeah I was thinking about that race last night and how to fix it. Http_read_request_bytes would not be canceled in this case. We can try adding a timeout on calling ReadToInputPipe(). Other than that, not sure what else to do besides synchronize.